edb/schema/pointers.py

```diff
@@ -664,8 +664,14 @@ def is_property(self, schema: s_schema.Schema) -> bool:
     def is_link_property(self, schema: s_schema.Schema) -> bool:
         raise NotImplementedError
 
-    def is_protected_pointer(self, schema: s_schema.Schema) -> bool:
-        return self.get_shortname(schema).name in {'id', '__type__'}
+    def is_protected_pointer(
+        self, exprtype: s_types.ExprType, schema: s_schema.Schema
+    ) -> bool:
+        sn = self.get_shortname(schema).name
+        return (
+            sn == '__type__'
+            or (sn == 'id' and not exprtype.is_mutation())
```

Comment:
Hm, I'd explicitly check for `Insert` here, because mutating `id` in `UPDATE` should remain prohibited.

Reply:
Do we have a test that it's prohibited? I think it'd make sense to add a few tests ensuring we have sane custom ID semantics.

Reply:
I made it `is_mutation` so that we get `cannot update property 'id': it is declared as read-only` message on update, which I thought was a better message. I can add a test for that, or I can change it

Reply:
Oh, I forgot that there's another check down the line that protects it. It's fine then.